### ISSUE TITLE
unevaluatedProperties, allOf and $ref

### ISSUE THREAD

**Issue by charlesgiroux** [gregsdennis/Manatee.Json#248]
My understanding of the 2019-09 draft was that unevaluatedProperties should see through $ref, but given the following schema:
```
{
    "definitions": {
        "other": {
            "type": "object",
            "properties": {
                "surfboard": {
                    "type": "string"
                }
            }
        }
    },
    "allOf": [
        {
            "$ref": "#/definitions/other"
        },
        {
            "properties": {
                "wheels": {},
                "headlights": {}
            }
        },
        {
            "properties": {
                "pontoons": {}
            }
        },
        {
            "properties": {
                "wings": {}
            }
        }
    ],
    "unevaluatedProperties": false
}
```
and the following json to be validated:
```
{ "pontoons": {}, "wheels": {}, "surfboard": "2" }
```
The following error is returned:
```
{
   "valid":false,
   "keywordLocation":"#/unevaluatedProperties",
   "instanceLocation":"#",
   "keyword":"unevaluatedProperties",
   "error":"Any properties not covered by `properties`, `patternProperties`, and `additionalProperties` failed validation."
}
```

I would expect the JSON to be valid, unless I'm wrong about "unevaluatedProperties" and "$ref"? I'm trying to compose parent schemas with subschemas  through the use of $ref and restrict the use of undefined properties.

I'm using v12.0.2.

**Comment by gregsdennis** [comment]
Thanks for the report.  I'll look into it.  Hm... I thought I had added the properties which cause the failure to the additional data in the results, but it doesn't look like I did.  I'll add that.

Also, this seems like it'd be a good candidate for https://github.com/json-schema-org/JSON-Schema-Test-Suite.

**Comment by charlesgiroux** [comment]
To clarify, did I understand the spec correctly?

Here is the Verbose error:

```
{
  "valid": false,
  "keywordLocation": "#",
  "instanceLocation": "#",
  "errors": [
    {
      "valid": true,
      "keywordLocation": "#/definitions",
      "instanceLocation": "#",
      "keyword": "definitions"
    },
    {
      "valid": true,
      "keywordLocation": "#/allOf",
      "instanceLocation": "#",
      "keyword": "allOf",
      "annotations": [
        {
          "valid": true,
          "keywordLocation": "#/allOf/0/$ref",
          "instanceLocation": "#",
          "keyword": "$ref",
          "annotations": [
            {
              "valid": true,
              "keywordLocation": "#/allOf/0/$ref",
              "instanceLocation": "#",
              "annotations": [
                {
                  "valid": true,
                  "keywordLocation": "#/allOf/0/$ref/type",
                  "instanceLocation": "#",
                  "keyword": "type"
                },
                {
                  "valid": true,
                  "keywordLocation": "#/allOf/0/$ref/properties",
                  "instanceLocation": "#",
                  "keyword": "properties",
                  "annotations": [
                    {
                      "valid": true,
                      "keywordLocation": "#/allOf/0/$ref/properties/surfboard",
                      "instanceLocation": "#/surfboard",
                      "annotations": [
                        {
                          "valid": true,
                          "keywordLocation": "#/allOf/0/$ref/properties/surfboard/type",
                          "instanceLocation": "#/surfboard",
                          "keyword": "type"
                        }
                      ]
                    }
                  ]
                }
              ]
            }
          ]
        },
        {
          "valid": true,
          "keywordLocation": "#/allOf/1",
          "instanceLocation": "#",
          "annotations": [
            {
              "valid": true,
              "keywordLocation": "#/allOf/1/properties",
              "instanceLocation": "#",
              "keyword": "properties",
              "annotations": [
                {
                  "valid": true,
                  "keywordLocation": "#/allOf/1/properties/wheels",
                  "instanceLocation": "#/wheels"
                }
              ]
            }
          ]
        },
        {
          "valid": true,
          "keywordLocation": "#/allOf/2",
          "instanceLocation": "#",
          "annotations": [
            {
              "valid": true,
              "keywordLocation": "#/allOf/2/properties",
              "instanceLocation": "#",
              "keyword": "properties",
              "annotations": [
                {
                  "valid": true,
                  "keywordLocation": "#/allOf/2/properties/pontoons",
                  "instanceLocation": "#/pontoons"
                }
              ]
            }
          ]
        },
        {
          "valid": true,
          "keywordLocation": "#/allOf/3",
          "instanceLocation": "#",
          "annotations": [
            {
              "valid": true,
              "keywordLocation": "#/allOf/3/properties",
              "instanceLocation": "#",
              "keyword": "properties"
            }
          ]
        }
      ]
    },
    {
      "valid": false,
      "keywordLocation": "#/unevaluatedProperties",
      "instanceLocation": "#",
      "keyword": "unevaluatedProperties",
      "error": "Any properties not covered by `properties`, `patternProperties`, and `additionalProperties` failed validation."
    }
  ]
}
```

**Comment by gregsdennis** [comment]
I think you're right. I may not be passing evaluated properties down through `$ref`.

@handrews?

**Comment by handrews** [comment]
@gregsdennis yeah looks like it.  @charlesgiroux is correct about the expected behavior.

The result of `$ref` should be the same as the result of the target schema.  That is true for both the validation result and for annotations.  I don't know if you're actually implementing `unevaluatedProperties` in terms of passing *up* annotations, but logically `surfboard` should appear as a `properties` annotation from the target schema.  It's totally valid to implement it some other way- I think the spec explicitly say so.

**Comment by gregsdennis** [comment]
Yeah, yeah. Up... Down... I'm in New Zealand, so it's just a different perspective.

**Comment by handrews** [comment]
> Yeah, yeah. Up... Down... I'm in New Zealand, so it's just a different perspective.

🙃

**Comment by gregsdennis** [comment]
@charlesgiroux this should be resolved in v12.0.3, out soon.
